payments/sap_invoices.py

```diff
@@ -0,0 +1,122 @@
+import contextlib
+import datetime
+from django.conf import settings
+from django.template import loader
+from django.utils import timezone
+from django.utils.translation import gettext as _
+
+from payments.utils import round_price
+from resources.models import Reservation
+
+
+@contextlib.contextmanager
+def get_reservations_for_invoicing():
+    """Returns reservations that are eligible for invoice generation.
+
+    This should include all CONFIRMED reservations marked invoice approved.
+
+    Will mark reservations as generated when done.
+    """
+    reservations = (
+        Reservation.objects.filter(
+            state=Reservation.CONFIRMED,
+            order__isnull=False,
+            invoice_requested=True,
+            invoice_approved_at__isnull=False,
+            invoice_generated_at__isnull=True,
+        )
+        .select_related(
+            "resource",
+            "resource__unit",
+            "order",
+        )
+        .prefetch_related(
+            "order__order_lines",
+            "order__order_lines__product",
+        )
+        .order_by("invoice_approved_at")
+    )
+
+    yield reservations
+
+    reservations.update(invoice_generated_at=timezone.now())
```

Comment:
I think this function should only be used for getting the eligible reservations, and then marking them as generated (`invoice_generated_at`) should happen elsewhere, e.g. inside `generate_sales_order` or in the management command. Of course the tests would need to be adjusted then too.

Reply:
Yes, that's fine, that probably makes more sense.